Mathlib/Analysis/Calculus/TangentCone.lean

```diff
@@ -207,6 +208,125 @@ theorem mem_tangentCone_of_segment_subset {s : Set G} {x y : G} (h : segment ℝ
+        · specialize hv n
+          simp only [mem_inter_iff, mem_diff, mem_singleton_iff, Metric.mem_ball, dist_eq_norm]
+            at hv
+          simpa using hv.2.le
+      _ = ((u n) ⁻¹ * u n ) * ‖r‖ * u n := by ring
+      _ = ‖r‖ * u n := by rw [inv_mul_cancel₀ (u_pos n).ne', one_mul]
+    refine ⟨?_, 0, fun n hn ↦ by simpa using c_ne n⟩
+    apply squeeze_zero (fun n ↦ by positivity) B
+    simpa using u_lim.const_mul _
+  refine ⟨c, d, Eventually.of_forall (fun n ↦ by simpa [d] using (hv n).1.1), c_lim, ?_⟩
+  rw [tendsto_zero_iff_norm_tendsto_zero]
+  exact squeeze_zero (fun n ↦ by positivity) (fun n ↦ (c_le n).le) u_lim
+
+/-- In a proper space, the tangent cone at a non-isolated point is nontrivial. -/
+theorem tangentCone_nonempty_of_properSpace [ProperSpace E]
+    {s : Set E} {x : E} (hx : (𝓝[s \ {x}] x).NeBot) :
+    (tangentConeAt 𝕜 s x ∩ {0}ᶜ).Nonempty := by
+  obtain ⟨u, -, u_pos, u_lim⟩ :
+      ∃ u, StrictAnti u ∧ (∀ (n : ℕ), 0 < u n) ∧ Tendsto u atTop (𝓝 (0 : ℝ)) :=
+    exists_seq_strictAnti_tendsto (0 : ℝ)
+  have A n : ((s \ {x}) ∩ Metric.ball x (u n)).Nonempty := by
+    apply NeBot.nonempty_of_mem hx (inter_mem_nhdsWithin _ (Metric.ball_mem_nhds _ (u_pos n)))
+  choose v hv using A
+  let d := fun n ↦ v n - x
+  have M n : x + d n ∈ s \ {x} := by simpa [d] using (hv n).1
+  let ⟨r, hr⟩ := exists_one_lt_norm 𝕜
+  have W n := rescale_to_shell hr zero_lt_one (x := d n) (by simpa using (M n).2)
+  choose c c_ne c_le le_c hc using W
+  have c_lim : Tendsto (fun n ↦ ‖c n‖) atTop atTop := by
+    suffices Tendsto (fun n ↦ ‖c n‖⁻¹ ⁻¹ ) atTop atTop by simpa
+    apply tendsto_inv_nhdsGT_zero.comp
+    simp only [nhdsWithin, tendsto_inf, tendsto_principal, mem_Ioi, norm_pos_iff, ne_eq,
+      eventually_atTop, ge_iff_le]
+    have B (n : ℕ) : ‖c n‖⁻¹ ≤ 1⁻¹ * ‖r‖ * u n := by
+      apply (hc n).trans
+      gcongr
+      specialize hv n
+      simp only [mem_inter_iff, mem_diff, mem_singleton_iff, Metric.mem_ball, dist_eq_norm] at hv
+      simpa using hv.2.le
+    refine ⟨?_, 0, fun n hn ↦ by simpa using c_ne n⟩
+    apply squeeze_zero (fun n ↦ by positivity) B
+    simpa using u_lim.const_mul _
+  obtain ⟨l, l_mem, φ, φ_strict, hφ⟩ :
+      ∃ l ∈ Metric.closedBall (0 : E) 1 \ Metric.ball (0 : E) (1 / ‖r‖),
+      ∃ (φ : ℕ → ℕ), StrictMono φ ∧ Tendsto ((fun n ↦ c n • d n) ∘ φ) atTop (𝓝 l) := by
+    apply IsCompact.tendsto_subseq _ (fun n ↦ ?_)
+    · exact (isCompact_closedBall 0 1).diff Metric.isOpen_ball
+    simp only [mem_diff, Metric.mem_closedBall, dist_zero_right, (c_le n).le,
+      Metric.mem_ball, not_lt, true_and, le_c n]
+  refine ⟨l, ?_, ?_⟩; swap
+  · simp only [mem_compl_iff, mem_singleton_iff]
+    contrapose! l_mem
+    simp only [one_div, l_mem, mem_diff, Metric.mem_closedBall, dist_self, zero_le_one,
+      Metric.mem_ball, inv_pos, norm_pos_iff, ne_eq, not_not, true_and]
+    contrapose! hr
+    simp [hr]
+  refine ⟨c ∘ φ, d ∘ φ, ?_, ?_, ?_⟩
```

Comment:
Would you like to inline at least the third term? (The other two are up to you.)